File: NSTViewWarmuper/WebViewController.m
//
//  WebViewController.m
//  NSTViewWarmuper
//
//  Created by Timur Bernikowich on 2/6/17.
//  Copyright © 2017 Timur Bernikovich. All rights reserved.
//

#import "WebViewController.h"
#import "NSTViewWarmuper.h"

@interface WebViewController () <WKNavigationDelegate, UIWebViewDelegate>

@property (nonatomic) UIWebView *UIWebView;
@property (nonatomic) WKWebView *WKWebView;

@property (nonatomic) NSTimeInterval loadHTMLStart;

@end

@implementation WebViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    if (self.webKit) {
        if (self.warmup) {
            self.WKWebView = [[NSTWKWebViewWarmuper sharedViewWarmuper] dequeueWarmupedWKWebView];
        } else {
            WKWebViewConfiguration *configuration = [WKWebViewConfiguration new];
            self.WKWebView = [[WKWebView alloc] initWithFrame:CGRectZero configuration:configuration];
        }
        self.WKWebView.scrollView.scrollEnabled = NO;
        self.WKWebView.frame = self.view.bounds;
        self.WKWebView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        self.WKWebView.navigationDelegate = self;
        [self.view addSubview:self.WKWebView];
    } else {
        if (self.warmup) {
            self.UIWebView = [[NSTUIWebViewWarmuper sharedViewWarmuper] dequeueWarmupedUIWebView];
        } else {
            self.UIWebView = [UIWebView new];
        }
        self.UIWebView.scrollView.scrollEnabled = NO;
        self.UIWebView.frame = self.view.bounds;
        self.UIWebView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        self.UIWebView.delegate = self;
        [self.view addSubview:self.UIWebView];
    }
    
    NSString *path = [[NSBundle mainBundle] pathForResource:@"some1" ofType:@"html" inDirectory:@""];
    NSString *contents = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:NULL];
    self.loadHTMLStart = CACurrentMediaTime();
    if (self.webKit) {
        [self.WKWebView loadHTMLString:contents baseURL:nil];
    } else {
        [self.UIWebView loadHTMLString:contents baseURL:nil];
    }
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation
{
    [self update];
}

- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    [self update];
}

- (void)update
{
    NSTimeInterval delta = CACurrentMediaTime() - self.loadHTMLStart;
    NSLog(@"END: %f, TIME: %f", CACurrentMediaTime(), delta);
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:[NSString stringWithFormat:@"%f", delta] style:UIBarButtonItemStylePlain target:self action:@selector(update)];
    self.navigationItem.rightBarButtonItem.enabled = NO;
}

- (void)dealloc
{
    self.WKWebView.navigationDelegate = nil;
    self.UIWebView.delegate = nil;
}

@end

File: Example/WebViewWarmUper/WebViewController.swift
//
//  Created by Timur Bernikovich on 07/03/2019.
//  Copyright © 2019 Timur Bernikovich. All rights reserved.
//

import UIKit
import WebKit
import WebViewWarmUper

enum WebViewType {
    case webKit
}

class WebViewController: UIViewController {
    
    private enum Resource: String {
        case rich = "apple"
        case articleWithWidgetAndCss = "sample1"
        case articleWithCss = "sample2"
    }
    
    let type: WebViewType
    let warmUp: Bool
    private var loadHTMLStart: TimeInterval = 0
    
    init(type: WebViewType, warmUp: Bool) {
        self.type = type
        self.warmUp = warmUp
        super.init(nibName: nil, bundle: nil)
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.backgroundColor = .white
        
        let webView: UIView & WebView
        switch type {
        case .webKit:
            let wkWebView: WKWebView
            if warmUp {
                wkWebView = WKWebViewWarmUper.shared.dequeue()
            } else {
                wkWebView = WKWebView(frame: .zero, configuration: WKWebViewConfiguration())
            }
            wkWebView.navigationDelegate = self
            wkWebView.scrollView.isScrollEnabled = false
            wkWebView.loadHTMLString("", baseURL: nil)
            webView = wkWebView
        }
        
        webView.frame = view.bounds
        webView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        view.addSubview(webView)
        
        let resource: Resource = .articleWithCss
        guard let path = Bundle.main.path(forResource: resource.rawValue, ofType: "html"),
            let contents = try? String(contentsOfFile: path) else {
            return
        }
        
        loadHTMLStart = CACurrentMediaTime()
        webView.loadHTMLString(contents)
    }
    
}

protocol WebView {
    func loadHTMLString(_ string: String)
}

extension WKWebView: WebView {
    func loadHTMLString(_ string: String) {
        loadHTMLString(string, baseURL: nil)
    }
}

extension WebViewController: WKNavigationDelegate {
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        updateResult()
    }
    private func updateResult() {
        let delta = CACurrentMediaTime() - loadHTMLStart
        navigationItem.rightBarButtonItem = UIBarButtonItem(title: String(format: "%.2f", delta), style: .plain, target: nil, action: nil)
        navigationItem.rightBarButtonItem?.isEnabled = false
    }
}
